README.md

```diff
@@ -14,18 +14,55 @@ Implement the tool `st` that does exact pattern matching using a suffix tree. Te
 
 ## Problems encountered if any.
+Other than getting one step closer to enlightenment, most problems we encountered were -as usual- temporary strokes of idiocy on our part (indexing -once again- is hard).
+Most difficulties we had were in early implemetation choices for our Knæ class, eg. do we want a dictionary or list implementation, and how do we handled it. How much information do we need in each Knæ, and how do we deal with our choice of ints as leaves. 
 
 ## Correctness
 
 *Describe experiments that verifies the correctness of your implementations.*
 
+We could not come up with a good idea for testing just the construction, but if the tree is constructed correctly the search should output the same matches as the naive implementation of exact pattern matching. 
+We generated test data and compared outputs from our ST implementation with the naive pattern matching, to verifiy correctness. Furthermore we also tried with empty strings, just in case something went wierd. For palindromes, markov, random and single letter, the output seemed correct. biggest problem was that it was not necessarily in the same order, but we don't care about that.
+
+
 ## Running time
 
 *Describe experiments that verifies that your implementation of `st` uses no more time than O(n) or O(n²) (depending on the algorithm) for constructing the suffix tree and no more than O(m) for searching for a given read in it. Remember to explain your choice of test data. What are “best” and “worst” case inputs?*
 
 *If you have graphs that show the running time--you probably should have--you can embed them here like we did in the previous project.*
 
+We generated 2 series of data for the running time, one series using increasing N (input string length), at three different choices of M (pattern length), the other keeping N constant and varying M. The relation between lines in the resulting plots should reveal the time complexity.
+
+When running the code on strings only consisting of 'A's with varying lengths of sequences we see a clear O(n^2) time complexity for the contruction of trees.
+
+![](figs/identical_contruct.png)
+
+The same time complexity is evident for a Markov-chain-generated string (see data_gen.py, DNA_markov() for generation related details), although it is more dependent of the sequence.
+
+![](figs/markov_construct.png)
+
+When looking at time complexities for the search algorithm the results are more obscure. 
+
+![](identical_search.png)
+
+The time complexity should be O(m) (m being pattern length), but it seems the short inputs run faster. We attribute this to the breath-first-traversal in our implimentation of the search algorithm. If the string is short we have to report many more occurences and thus go though more nodes to get there. The same pattern is seen for a markov-chain-generated sequence as shown below.
```

Comment:
Well, here's the thing about $O(m)$: shorter (smaller $m$) _does_ mean faster!

What the figure shows, of course, is that longer sequences run faster, and there is a good reason for that. A reason that has nothing to do with breadth-first traversal; it is traversal in general. The running time _isn't_ $O(m)$ for a search, it is $O(n+z)$ as you also observe, and with shorter sequences, you are right again, you have more occurrences. In other words, $z$ will go up as $m$ goes down.

The question then is what to do about that when measuring performance?

The answer is as easy as posing the question: split the measurements into two so you measure the two phases of the search separately. You can check if searching down the tree takes time $O(m)$, and that traversal takes time $O(z)$. All it takes is two different clocks (or, rather, two pairs of start/stop integers).

For $m$, you know ahead of time what it is, but for $z$, you need to collect that while you do the traversal; otherwise, you cannot plot time versus $z$. But it isn't really that hard to count.